File: backend/internal/model/budget.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
	"gorm.io/gorm"
)

// Budget is a per-category spend limit per period, owned by exactly one of a
// user (personal) or a household (shared) — ADR-0018. The DB CHECK
// budgets_owner_chk enforces the XOR; repository.PlanOwner is the data-layer
// counterpart. Exactly one of UserID / HouseholdID is non-nil.
type Budget struct {
	ID          int64           `gorm:"primaryKey" json:"id"`
	UserID      *int64          `json:"user_id"`
	HouseholdID *int64          `json:"household_id"`
	CategoryID  int64           `gorm:"not null" json:"category_id"`
	Period      string          `gorm:"not null" json:"period"`
	Amount      decimal.Decimal `gorm:"type:numeric(30,18);not null" json:"amount"`
	Rollover    bool            `gorm:"not null;default:false" json:"rollover"`
	// IsActive: the DB column has DEFAULT TRUE (migration 000001), but we
	// deliberately leave the GORM tag without `default:true` so explicit
	// `false` from the service layer survives INSERT — gorm omits zero-
	// valued fields when the tag carries a default clause.
	IsActive  bool           `gorm:"not null" json:"is_active"`
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	Category *Category `gorm:"foreignKey:CategoryID" json:"-"`
}

func (Budget) TableName() string { return "budgets" }

File: frontend/src/types/budget.ts
// Mirrors backend model.Budget + service.BudgetSpend.
export const BUDGET_PERIODS = ['monthly', 'weekly', 'annual'] as const
export type BudgetPeriod = (typeof BUDGET_PERIODS)[number]

export type Budget = {
  id: number
  // Owned by exactly one of user_id (personal) / household_id (ADR-0018).
  // Personal-scope responses from this client always carry a user_id.
  user_id: number
  household_id: number | null
  category_id: number
  period: BudgetPeriod
  amount: string
  rollover: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export type BudgetSpend = {
  budget_id: number
  category_id: number
  period: BudgetPeriod
  period_start: string
  period_end: string
  limit: string
  spent: string
  remaining: string
  pct: number
}

export type CreateBudgetInput = {
  category_id: number
  period: BudgetPeriod
  amount: string
  rollover?: boolean
  is_active?: boolean
}

export type UpdateBudgetInput = {
  category_id?: number
  period?: BudgetPeriod
  amount?: string
  rollover?: boolean
  is_active?: boolean
}
